docs/specifications/tests/Zone-TP/zone01.md

```diff
@@ -3,43 +3,209 @@
+
+The value in the Level column is the default severity level of the message. The
+severity level can be changed in the [Zonemaster-Engine profile]. Also see the
+[Severity Level Definitions] document.
+
+The argument names in the Arguments column lists the arguments used in the
+message. The argument names are defined in the [argument list].
+
+## Test procedure
+
+In this section and unless otherwise specified below, the terms "[DNS Query]"
+follow the specification for DNS queries as specified in [DNS Query and Response Defaults].
+The handling of the DNS responses on the DNS queries follow, unless otherwise specified below,
+what is specified for [DNS Response] in the same specification.
+
+1. Create a [DNS Query] with query type SOA and query name *Child Zone* ("SOA Query").
+
+2. Create the following empty sets:
+   1. Name server IP address, name, SERIAL ("MNAME Nameservers")
+   2. Name server IP address, SERIAL ("SERIAL Nameservers")
+   3. Name server IP address ("MNAME Not Master")
+   4. Name server IP address ("MNAME Is Master")
+   5. Name server IP address ("MNAME Is Localhost")
+   6. Name server IP address ("MNAME Is Dot")
+
+3. Obtain the set of name server IP addresses using [Method4] and [Method5] ("Name Server IP").
+
+4. For each name server IP address in *Name Server IP* do:
+   1. Send *SOA Query* to the name server and collect the response.
+   2. Go to next name server IP address if at least one of the following criteria is met:
+      1. There is no DNS response.
+      2. [RCODE Name] of the response is not "NoError".
+      3. The AA flag is not set in the response.
+      4. There is no SOA record with owner name matching the query.
+   3. From the DNS response, extract the name server name from the MNAME field:
+      1. If the name is "localhost", then add name server IP address to the *MNAME Is Localhost* set and go to next name server.
+      2. Else if the name is ".", then add name server IP address to the *MNAME Is Dot* set and go to next name server.
+      3. Else, add name server name to the *MNAME Nameservers* set.
+   4. From the DNS response, add the SERIAL field and name server IP address to the *SERIAL Nameservers* set.
+   5. Go to next name server.
+
+5. If the set *MNAME Is Localhost* is non-empty, then output *[Z01_MNAME_IS_LOCALHOST]* with name servers IP address(es).
+
+6. If the set *MNAME Is Dot* is non-empty, then output *[Z01_MNAME_IS_DOT]* with name servers IP address(es).
+
+7. If the set *MNAME Nameservers* is empty, then terminate these procedures.
+
+8. Else, for each name server name in *MNAME Nameservers* do:
+   1. Do a [DNS Lookup] of name server name (A and AAAA) and add the name server IP address(es) found to the *MNAME Nameservers* set.
+   2. If at least one IP address from the previous step was found, then:
+      1. For each name server IP for the name server name do: 
```

Comment:
Should there be exceptions for IP addresses 127.0.0.1 and ::1? I.e. another message and no query?

Reply:
The localhost IP addresses are special. The only "foreign" servers that Zonemaster connects to are name servers. It does no check on parent or grand-parent name servers when it comes to IP address. The name servers from delegation or the zone of *Child Zone* there is one test case that verifies the IP address and reacts on localhost address (among other non-public or non-global addresses). I do not think that this test case should repeat the entire IP address check, but it could make sense to capture the localhost addresses by having a special message and special code:

Message Tag                   | Level   | Arguments             | Message ID for message tag
:---------------------------- |:--------|:----------------------|:---------------------------------------------------------------------------------------
Z01_MNAME_HAS_LOCALHOST_ADDR      | WARNING   | nsname, ns_ip  | SOA MNAME name server "{nsname}" resolves to a localhost IP address ({ns_ip}).

Reply:
Sure, added.